workers/src/chain/contracts.rs

```diff
@@ -0,0 +1,129 @@
+//! Utilities for interacting with onchain contracts.
+
+use crate::chain::{ContractInst, HttpProvider};
+use alloy::{
+    contract::{ContractInstance, Interface},
+    dyn_abi::DynSolValue,
+    network::Ethereum,
+    primitives::{keccak256, Address, U256},
+    transports::http::{Client, Http},
+};
+use alloy_json_abi::JsonAbi;
+use eyre::{eyre, OptionExt, Result};
+use tracing::{debug, error};
+
+/// Create a contract instance from the ABI to interact with on-chain instance.
+pub fn create_contract_instance(addr: Address, http_provider: HttpProvider, abi: JsonAbi) -> Result<ContractInst> {
+    let contract: ContractInstance<Http<Client>, _, Ethereum> =
+        ContractInstance::new(addr, http_provider.clone(), Interface::new(abi));
+    Ok(contract)
+}
+
+/// Get the address of the MessageLib on the destination chain
+pub async fn get_messagelib_addr(contract: &ContractInst, eid: U256) -> Result<Address> {
+    // Call the `getUlnConfig` function on the contract
+    let receive_library = contract
+        .function(
+            "getReceiveLibrary",
+            &[DynSolValue::Address(*contract.address()), DynSolValue::Uint(eid, 32)],
+        )?
+        .call()
+        .await?;
+
+    match receive_library[0] {
+        DynSolValue::Address(address) => Ok(address),
+        _ => {
+            error!("Failed to get address");
+            Err(eyre!("Failed to get address"))
+        }
+    }
+}
+
+/// Get the number of required confirmations by the ULN.
+pub async fn query_confirmations(contract: &ContractInst, eid: U256) -> Result<U256> {
+    // Call the `getUlnConfig` function on the contract
+    let uln_config = contract
+        .function(
+            "getUlnConfig",
+            &[DynSolValue::Address(*contract.address()), DynSolValue::Uint(eid, 32)],
+        )?
+        .call()
+        .await?;
+
+    match &uln_config[0] {
```

Comment:
```suggestion
    match &uln_config.first()? {
```

Are we certain there's one config here and the first is the one we need?

Reply:
Everytime I tested, I got only one. I assumed it returns the config (a Tuple value) wrapped in an array.

Reply:
Yes, so lets make sure that the codebase is not using unsafe rust incase some random API stops returning what we have _**assumed**_